### PR TITLE
Fastroute with make wire parasitics

### PR DESCRIPTION
**Pull request by arthurjolo**
Fixes [#3039](https://github.com/The-OpenROAD-Project/OpenROAD/issues/3039).

Currently the slack is estimated on the beginning of global routing, and never considers the changes made during the fastroute that can make a non-critical net to become one with a large detour from their original routing.
This changes update the slack estimates during the fastroute when the critical_nets_percentage is set, also for the nets in the critical percentage if they have a large detour from their original routing and weren't chosen for a rip up, decrease their search area and select them to be re-routed so they have another chance to get a routing closer to their original.

### CODE REVIEW

**Comment by QuantamHD** [comment]
It feels like this option should be on by default, and have an option to turn it off.

**Comment by arthurjolo** [comment]
> It feels like this option should be on by default, and have an option to turn it off.

Currently the slack is only considered if the critical_nets_percentage is set, so I decided to keep this property, but it culd be added a default percentage.

File: src/grt/include/grt/Grid.h
Comment (maliberty) [inline src/grt/include/grt/Grid.h:1]: I'm unclear why this and other private headers have been made public.  I don't see a change outside of grt requiring this.

Comment (arthurjolo) [reply src/grt/include/grt/Grid.h:1]: Some of them are needed to include the MakeWireParasitics.h in fastroute, but at the time I just made them all public and forgot to put the ones that are not needed back. I am going to look again and see witch ones can go back to private.

Comment (maliberty) [reply src/grt/include/grt/Grid.h:1]: The public include area is for clients outside of grt.  This is really about internal includes.  I would make an abstract interface in fastroute/include which MakeWireParasitics can implement in order to keep the layering and avoid having fastroute need to include from grt.

File: src/grt/src/fastroute/CMakeLists.txt [hunk 1]
```diff
@@ -58,5 +58,6 @@ target_link_libraries(FastRoute4.1
     utl_lib
     stt_lib
     odb
+    dbSta_lib
```
Comment (maliberty) [inline src/grt/src/fastroute/CMakeLists.txt:61]: Why is this needed?

File: src/grt/src/MakeWireParasitics.cpp
Comment (maliberty) [inline src/grt/src/MakeWireParasitics.cpp:1]: Is there any algorithmic change here or is this just moving code around?

File: src/grt/src/GlobalRouter.cpp [hunk 2]
```diff
@@ -429,6 +429,8 @@ NetRouteMap GlobalRouter::findRouting(std::vector<Net*>& nets,
 {
   NetRouteMap routes;
   if (!nets.empty()) {
+    MakeWireParasitics builder(logger_, resizer_, sta_, db_->getTech(), this);
+    fastroute_->setMakeWireParasiticsBuilder(&builder);
     routes = fastroute_->run();
```
Comment (maliberty) [inline src/grt/src/GlobalRouter.cpp:434]: It would be good to "fastroute_->setMakeWireParasiticsBuilder(nullptr)" after run() to ensure the builder isn't used after it goes out of scope.

File: src/grt/src/fastroute/src/FastRoute.cpp [hunk 3]
```diff
@@ -682,6 +684,98 @@ NetRouteMap FastRouteCore::getRoutes()
   return routes;
 }
 
+NetRouteMap FastRouteCore::getPartialRoutes()
+{
+  NetRouteMap routes;
+  bool is_vertical = ((2 % 2) - layer_orientation_) != 0;
```
Comment (maliberty) [inline src/grt/src/fastroute/src/FastRoute.cpp:690]: 2 % 2 ??  Its a constant...

File: src/grt/src/fastroute/src/FastRoute.cpp [hunk 4]
```diff
@@ -682,6 +684,98 @@ NetRouteMap FastRouteCore::getRoutes()
   return routes;
 }
 
+NetRouteMap FastRouteCore::getPartialRoutes()
```
Comment (maliberty) [inline src/grt/src/fastroute/src/FastRoute.cpp:687]: What is the definition of a partial route?

File: src/grt/src/fastroute/src/FastRoute.cpp [hunk 5]
```diff
@@ -682,6 +684,98 @@ NetRouteMap FastRouteCore::getRoutes()
...
+        const std::vector<short>& gridsY = treeedge->route.gridsY;
+        const std::vector<short>& gridsL = treeedge->route.gridsL;
+        int lastX = tile_size_ * (gridsX[0] + 0.5) + x_corner_;
+        int lastY = tile_size_ * (gridsY[0] + 0.5) + y_corner_;
+        int lastL_h;
+        int lastL_v;
+        if (gridsL.empty()) {
+          if (is_vertical) {
+            lastL_h = 1;
+            lastL_v = 2;
```
Comment (maliberty) [inline src/grt/src/fastroute/src/FastRoute.cpp:715]: What are 1 & 2?  Are these directions?

File: src/grt/src/fastroute/src/RipUp.cpp [hunk 6]
```diff
@@ -326,6 +326,71 @@ bool FastRouteCore::newRipupCheck(const TreeEdge* treeedge,
   }
 }
 
+bool FastRouteCore::newRipupCheck(const TreeEdge* treeedge,
```
Comment (maliberty) [inline src/grt/src/fastroute/src/RipUp.cpp:329]: This looks to be a copy of the existing newRipupCheck.  Is it possible to avoid duplicating the code?

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/grt/src/fastroute/CMakeLists.txt (hunk 1 above)
Comment (arthurjolo) [reply src/grt/src/fastroute/CMakeLists.txt:61]: It is needed to include the definition of slack, but if needed I can change it to be just float instead.

File: src/grt/src/MakeWireParasitics.cpp
Comment (arthurjolo) [reply src/grt/src/MakeWireParasitics.cpp:1]: Since your last rewiew just mooved some code, but in general there is a new function to estimate parasitics from information that fastroute has.

Comment (arthurjolo) [reply src/grt/src/MakeWireParasitics.cpp:1]: But mostly it is the same code

File: src/grt/src/fastroute/src/FastRoute.cpp (hunk 3 above)
Comment (arthurjolo) [reply src/grt/src/fastroute/src/FastRoute.cpp:690]: Yes, It can be one

File: src/grt/src/fastroute/src/FastRoute.cpp (hunk 5 above)
Comment (arthurjolo) [reply src/grt/src/fastroute/src/FastRoute.cpp:715]: These are the layers definition for vertical edges and horizontal edges

File: src/grt/src/fastroute/src/RipUp.cpp (hunk 6 above)
Comment (arthurjolo) [reply src/grt/src/fastroute/src/RipUp.cpp:329]: Yes it can be one function, I created a new one because this fuction is used in other places than mazeRouteMSMD, and it should the added case should't be analized outside it, and at the time there was no way to difference them, but now there is, and I can change it.

File: src/grt/src/fastroute/src/FastRoute.cpp (hunk 4 above)
Comment (arthurjolo) [reply src/grt/src/fastroute/src/FastRoute.cpp:687]: I think partial route was a bad naming, but it is the routes before layer assignment.

Comment (arthurjolo) [reply src/grt/src/fastroute/src/FastRoute.cpp:687]: I was trying to make it possible to use both before and after layer assignment, then it could be just one function with getRoutes(),  if this makes more sense I can do it.

**Comment by maliberty** [comment]
> There is a public design failing and it is on investigation.

Is this resolved?  Has this been through secure CI?  Is there a default percentage?

**Comment by arthurjolo** [comment]
> > There is a public design failing and it is on investigation.
> 
> Is this resolved? Has this been through secure CI? Is there a default percentage?

The public design failing was resolved, and it has been through a secure CI, I am waiting for @eder-matheus report on this run. As a default perecentege I didn't look that much into it, but for the test case attached to the issue 5% of critical nets seemd to be the best alternative, the secure run also used 5%.

**Comment by eder-matheus** [comment]
> > > There is a public design failing and it is on investigation.
> > 
> > 
> > Is this resolved? Has this been through secure CI? Is there a default percentage?
> 
> The public design failing was resolved, and it has been through a secure CI, I am waiting for @eder-matheus report on this run. As a default perecentege I didn't look that much into it, but for the test case attached to the issue 5% of critical nets seemd to be the best alternative, the secure run also used 5%.

We actually have 2 separate runs in the CI. When using the default percentage of 0% (i.e., the updates from this PR are not executed), public and private CI are passing.

When using 5%, we still have some metric failures.

From the CI side, I'm fine with merging this PR and do the fix for the matric failures with 5% in a future PR.

**Comment by maliberty** [comment]
> From the CI side, I'm fine with merging this PR and do the fix for the matric failures with 5% in a future PR.

agreed

File: src/grt/src/fastroute/src/FastRoute.cpp (hunk 4 above)
Comment (maliberty) [reply src/grt/src/fastroute/src/FastRoute.cpp:687]: If it can convert both sorts of routes that is good.  If not we should rename it.

Comment (arthurjolo) [reply src/grt/src/fastroute/src/FastRoute.cpp:687]: Oh it seems it wasn’t pushed, but I made both methods into one. I will push it.

File: src/grt/src/fastroute/src/FastRoute.cpp [hunk 7]
```diff
@@ -706,61 +708,53 @@ NetRouteMap FastRouteCore::getPartialRoutes()
...
-          int second_x = tile_size_ * (gridsX[1] + 0.5) + x_corner_;
-          lastL = (lastX == second_x) ? layer_v : layer_h;
+
+        if (layer_orientation_ != 0) {
+          layer_h = 1;
+          layer_v = 2;
         } else {
-          lastL = gridsL[0];
+          layer_h = 1;
+          layer_v = 2;
```
Comment (rovinski) [inline src/grt/src/fastroute/src/FastRoute.cpp:722]: If and else block are the same?

File: src/grt/src/fastroute/src/FastRoute.cpp (hunk 5 above)
Comment (arthurjolo) [reply src/grt/src/fastroute/src/FastRoute.cpp]: lastL_h and lastL_v were renamed to layer_h and layer_v respectively.

File: src/grt/src/fastroute/src/FastRoute.cpp (hunk 4 above)
Comment (arthurjolo) [reply src/grt/src/fastroute/src/FastRoute.cpp]: I tried to merge this method with the method getRoutes(), but this caused some errors in other parts of the flow. So this method was renamed to getPlanarRoutes(), for better understanding of it's purpose: generate the routes before layer assignment.

File: src/grt/src/MakeWireParasitics.cpp [hunk 8]
```diff
@@ -307,6 +367,111 @@ void MakeWireParasitics::makeParasiticsToPin(
...
+    auto ap_position = ap_positions.front();
+    pt = ap_position.first;
+    grid_pt = ap_position.second;
+  }
+  if (pin.isConnectedToPadOrMacro() || pin.isPort()) {
+    grid_pt = grouter_->findFakePinPosition(pin, net);
+  }
+  // Use the route layer above the pin layer if there is a via
+  // to the pin.
+  int layer = 3;
```
Comment (maliberty) [inline src/grt/src/MakeWireParasitics.cpp:412]: Why 3?

Comment (arthurjolo) [reply src/grt/src/MakeWireParasitics.cpp:412]: I define the layers for planar routes as m2 or m3.

Comment (arthurjolo) [reply src/grt/src/MakeWireParasitics.cpp:412]: with layer_h and layer_v beeing 1 (m2) or 2 (m3).

Comment (arthurjolo) [reply src/grt/src/MakeWireParasitics.cpp:412]: And it had to be hardcoded because the it wasn't possible to use the pin access point before layer assignment.

**Review by maliberty — APPROVED** [review]
